pydomcfg/domzgr/zgr.py

```diff
@@ -0,0 +1,164 @@
+                 sigma =  0 at the shallower boundary
+                 sigma = -1 at the deeper boundary
+
+        Parameters
+        ----------
+        k: int
+            Model level index. Note that
+            *) T-points are at integer values (between 1 and jpk)
+            *) W-points are at integer values - 1/2 (between 0.5 and jpk-0.5)
+        grd: str
+            If we are dealing with "T" or "W" model levels.
+
+        Returns
+        -------
+        ps: float
+            Uniform non-dimensional sigma-coordinate (0. <= sigma <= -1)
+        """
+
+        kindx = float(k + 1)  # to deal with python convention
+        if grd == "W":
+            kindx -= 0.5
+
+        ps = -(kindx - 0.5) / float(self._jpk - 1)
+        return ps
+
+    # -------------------------------------------------------------------------
+    @staticmethod
+    def compute_z3(
+        su: float,
+        ss1: float,
+        a1: float,
+        a2: float,
+        a3: float,
+        ss2: float = 0.0,
+        a4: float = 0.0,
+    ):
+        """
+        Generalised function providing the analytical
+        transformation from computational space to
+        physical space. It takes advantage of the fact that
+        z-coordinates can be considered a special case of
+        s-coordinate.
+
+        Note that z is downward positive.
+
+        Parameters
+        ----------
+        su: float
+            uniform non-dimensional vertical coordinate s,
+            aka sigma-coordinates. 0 <= s <= 1
+        ss1: float
+            stretched non-dimensional vertical coordinate s,
+            0 <= s <= 1
+        a1: float
+            parameter of the transformation
+        a2: float
+            parameter of the transformation
+        a3: float
+            parameter of the transformation
+        ss2: float
+            second stretched non-dimensional vertical coordinate s,
+            0 <= s <= 1 (only used for zco with ldbletanh = True)
+        a4: float
+            parameter of the transformation (only used for zco with ldbletanh = True)
+
+        Returns
+        -------
+        z: float
+            Depths of model levels
+        """
+
+        z = a1 + a2 * su + a3 * ss1 + a4 * ss2
+        return z
+
+    # -------------------------------------------------------------------------
+    def compute_e3(self, ds: Dataset):
+        """
+        Grid cell thickness computed as discrete derivative
+        (central-difference) of levels' depth
+
+        Parameters
+        ----------
+        ds: Dataset
+            xarray dataset with empty ``e3{T,W}`` and ``z3{T,W}``
+            correctly computed
+        Returns
+        -------
+        ds: Dataset
+            xarray dataset with ``e3{T,W}`` correctly computed
+        """
+        for k in range(self._jpk - 1):
+            ds["e3T"][{"z": k}] = ds["z3W"][{"z": k + 1}] - ds["z3W"][{"z": k}]
+            ds["e3W"][{"z": k + 1}] = ds["z3T"][{"z": k + 1}] - ds["z3T"][{"z": k}]
```

Comment:
Alternatively, you could use pre-built `diff` rather than building yourself the for loop.
```suggestion
        ds["e3T"][{"z": slice(None, -1)}] = ds["z3W"].diff("z", label="lower")
        ds["e3W"][{"z": slice(1, None)}] = ds["z3T"].diff("z", label="upper")
```

Reply:
It is not very clear to me the meaning of 'upper' and 'lower' ... also in the doc it is not very clear I think (http://xarray.pydata.org/en/stable/generated/xarray.DataArray.diff.html)  ... a part from that yes, I agree with your suggestions, although I think we should keep in mind also code readability - sometimes it is cleaner, clearer and easier to mantain/debug using a simpler syntax ...

Reply:
Yes good point about readability! In general I think selecting using whole slices gives better performance, although I'm not sure whether it applies here.

Diff creates an array of size `jpk - 1`. Label lower and upper sets the resulting coord and index starting at 0 or 1, respectively. (You can probably always use the default if `z` is dimension but not coord/index, e.g. if you initialize doing `.expand_dims(z=jpk)` rather than `.expand_dims(z=range(jpk))`)

Here is an example:
```python
import xarray as xr
da = xr.DataArray(range(10), dims="dim_0", coords={"dim_0": range(10)})
print(da)
for label in ["lower", "upper"]:
    print('\n'+label)
    print(da.diff("dim_0", label=label).coords)
```
```
<xarray.DataArray (dim_0: 10)>
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
Coordinates:
  * dim_0    (dim_0) int64 0 1 2 3 4 5 6 7 8 9

lower
Coordinates:
  * dim_0    (dim_0) int64 0 1 2 3 4 5 6 7 8

upper
Coordinates:
  * dim_0    (dim_0) int64 1 2 3 4 5 6 7 8 9

```

Reply:
If `z` is coordinate and you don't use `label` in `diff`, then when you try to substitute the values `xarray` checks the coordinates and raises an error if there is a mismatch.

Reply:
I see, thanks for the explanation!